contracts/token/ERC1155/ERC1155.sol

```diff
@@ -143,44 +143,75 @@ contract ERC1155 is Context, ERC165, IERC1155, IERC1155MetadataURI {
     }
 
     /**
-     * @dev Transfers `amount` tokens of token type `id` from `from` to `to`.
+     * @dev Transfers `amount` tokens of token type `id` from `from` to `to`. Will mint (or burn) if `from` (or `to`) is the zero address.
      *
-     * Emits a {TransferSingle} event.
+     * Emits a {TransferSingle} event if only one transfer is done, and {TransferBatch} event for multiple transfer operations.
      *
      * Requirements:
      *
-     * - `to` cannot be the zero address.
-     * - `from` must have a balance of tokens of type `id` of at least `amount`.
-     * - If `to` refers to a smart contract, it must implement {IERC1155Receiver-onERC1155Received} and return the
+     * - If `to` refers to a smart contract, it must implement {IERC1155Receiver-onERC1155BatchReceived} and return the
      * acceptance magic value.
      */
-    function _safeTransferFrom(
+    function _update(
         address from,
         address to,
-        uint256 id,
-        uint256 amount,
+        uint256[] memory ids,
+        uint256[] memory amounts,
         bytes memory data
     ) internal virtual {
-        require(to != address(0), "ERC1155: transfer to the zero address");
+        require(ids.length == amounts.length, "ERC1155: ids and amounts length mismatch");
 
         address operator = _msgSender();
-        uint256[] memory ids = _asSingletonArray(id);
-        uint256[] memory amounts = _asSingletonArray(amount);
 
-        _beforeTokenTransfer(operator, from, to, ids, amounts, data);
-
-        uint256 fromBalance = _balances[id][from];
-        require(fromBalance >= amount, "ERC1155: insufficient balance for transfer");
-        unchecked {
-            _balances[id][from] = fromBalance - amount;
-        }
-        _balances[id][to] += amount;
+        for (uint256 i = 0; i < ids.length; ++i) {
+            uint256 id = ids[i];
+            uint256 amount = amounts[i];
```

Comment:
We can define and use a function `_unsafeAccess` here since we don't need bounds checking.

We could add this function in the `Arrays` library. Previously we insisted in keeping this function private, but I think as long as it is called "unsafe" we can make it available. What do you think @Amxx?

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/d59306bd06a241083841c2e4a39db08e1f3722cc/contracts/utils/Checkpoints.sol#L199-L207

Reply:
`unsafeAccess`'s main point is to avoid the `sload` that gets the length. Here, we are in memory, so we would avoid a `mload` (and a comparison). It would save gas, but not as much as when dealing with storage.

Its really a balance of savings vs readability.

I'm ok with implementing it ... maybe in a following PR were we would go through the codebase and finds all relevant accesses.

Reply:
That's true, the savings are a lot less significant in memory. But in relative terms doing bounds checking probably doubles or triples the cost of an array access. We should look into this, but agree that it's not critical.